File: src/models/dex.js
'use strict';

const Bookshelf = require('../libraries/bookshelf');
const Game      = require('./game');
const Knex      = require('../libraries/knex');

module.exports = Bookshelf.model('Dex', Bookshelf.Model.extend({
  tableName: 'dexes',
  hasTimestamps: ['date_created', 'date_modified'],
  game () {
    return this.belongsTo(Game, 'game_id');
  },
  caught () {
    return Knex('captures').count().where('dex_id', this.get('id'))
    .then((res) => parseInt(res[0].count));
  },
  virtuals: {
    total () {
      if (this.get('regional')) {
        return this.related('game').related('game_family').get('regional_total');
      }

      return this.related('game').related('game_family').get('national_total');
    }
  },
  serialize () {
    return this.caught()
    .then((caught) => {
      return {
        id: this.get('id'),
        user_id: this.get('user_id'),
        title: this.get('title'),
        slug: this.get('slug'),
        shiny: this.get('shiny'),
        game: this.related('game').serialize(),
        regional: this.get('regional'),
        caught,
        total: this.get('total'),
        date_created: this.get('date_created'),
        date_modified: this.get('date_modified')
      };
    });
  }
}, {
  RELATED: ['game', 'game.game_family']
}));

File: pkg/models/dex.go
package models

import "encoding/json"

// Dex is a model representing a dex.
type Dex struct {
	tableName struct{} `sql:"dexes,alias:dexes"`

	ID       int    `json:"id"`
	UserID   int    `json:"user_id"`
	Title    string `json:"title"`
	Slug     string `json:"slug"`
	Shiny    bool   `sql:",notnull" json:"shiny"`
	GameID   string `json:"-"`
	Game     Game   `json:"game"`
	Regional bool   `sql:",notnull" json:"regional"`
}

// MarshalJSON is used to satisfy the Marshaler interface to customize the way
// it's serialized into JSON.
func (d Dex) MarshalJSON() ([]byte, error) {
	total := d.Game.GameFamily.NationalTotal
	if d.Regional {
		total = d.Game.GameFamily.RegionalTotal
	}

	type alias Dex
	return json.Marshal(&struct {
		alias
		Caught int `json:"caught"` // TODO: actually get caught amount
		Total  int `json:"total"`
	}{
		alias: (alias)(d),
		Total: total,
	})
}
